homeassistant/components/edp_redy.py

```diff
@@ -0,0 +1,340 @@
+"""
+Support for EDP re:dy.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/edp_redy/
+"""
+
+import asyncio
+import json
+import logging
+
+import aiohttp
+import async_timeout
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.const import (CONF_USERNAME, CONF_PASSWORD,
+                                 EVENT_HOMEASSISTANT_START)
+from homeassistant.core import callback
+from homeassistant.helpers import discovery, dispatcher, aiohttp_client
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_point_in_time
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'edp_redy'
+EDP_REDY = "edp_redy"
+DATA_UPDATE_TOPIC = '{0}_data_update'.format(DOMAIN)
+ACTIVE_POWER_ID = "home_active_power"
+
+URL_BASE = "https://redy.edp.pt/EdpPortal/"
+URL_LOGIN_PAGE = URL_BASE
+URL_GET_ACTIVE_POWER = "{0}/Consumption/GetActivePower".format(URL_BASE)
+URL_GET_SWITCH_MODULES = "{0}/HomeAutomation/GetSwitchModules".format(URL_BASE)
+URL_SET_STATE_VAR = "{0}/HomeAutomation/SetStateVar".format(URL_BASE)
+URL_LOGOUT = "{0}/Login/Logout".format(URL_BASE)
+
+UPDATE_INTERVAL = 30
+DEFAULT_TIMEOUT = 30
+SESSION_TIME = 59
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+class EdpRedySession:
+    """Representation of an http session to the service."""
+
+    def __init__(self, hass, username, password):
+        """Init the session."""
+        self._username = username
+        self._password = password
+        self._session = None
+        self._session_time = dt_util.utcnow()
+        self._hass = hass
+        self.modules_dict = {}
+        self.values_dict = {}
+
+    async def async_init_session(self):
+        """Create a new http session."""
+        payload_auth = {'username': self._username,
+                        'password': self._password,
+                        'screenWidth': '1920', 'screenHeight': '1080'}
+
+        try:
+            # create session and fetch login page
+            session = aiohttp_client.async_get_clientsession(self._hass)
+            with async_timeout.timeout(DEFAULT_TIMEOUT, loop=self._hass.loop):
+                resp = await session.get(URL_LOGIN_PAGE)
```

Comment:
Device specific I/O and data parsing should be moved to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Reply:
Even something as small as this? Ok.